sw/src/lib/registration.js

```diff
@@ -0,0 +1,164 @@
+            // serve assets; see https://github.com/facebook/create-react-app/issues/2374
+            return
+        }
+
+        window.addEventListener('load', () => {
+            // By compiling the dev SW to the 'public' dir, this URL works in
+            // both dev and production modes
+            const swUrl = `${process.env.PUBLIC_URL}/service-worker.js`
+
+            if (isLocalhost) {
+                // This is running on localhost. Let's check if a service worker still exists or not.
+                checkValidServiceWorker(swUrl, config)
+
+                // Add some additional logging to localhost, pointing developers
+                // to the service worker/PWA documentation.
+                navigator.serviceWorker.ready.then(() => {
+                    console.log(
+                        'This web app is using a service worker. If this is ' +
+                            'a production environment, it is being served ' +
+                            'cache-first. Learn more at https://cra.link/PWA'
+                    )
+                })
+            } else {
+                // Is not localhost. Just register service worker
+                registerValidSW(swUrl, config)
+            }
+        })
+    }
+}
+
+function registerValidSW(swUrl /* config */) {
+    navigator.serviceWorker.register(swUrl).catch(error => {
+        console.error('Error during service worker registration:', error)
+    })
+}
+
+function checkValidServiceWorker(swUrl, config) {
+    // Check if the service worker can be found. If it can't reload the page.
+    fetch(swUrl, {
+        headers: { 'Service-Worker': 'script' },
+    })
+        .then(response => {
+            // Ensure service worker exists, and that we really are getting a JS file.
+            const contentType = response.headers.get('content-type')
+            if (
+                response.status === 404 ||
+                (contentType != null &&
+                    contentType.indexOf('javascript') === -1)
+            ) {
+                // No service worker found. Probably a different app. Reload the page.
+                navigator.serviceWorker.ready.then(registration => {
+                    registration.unregister().then(() => {
+                        window.location.reload()
+                    })
+                })
+            } else {
+                // Service worker found. Proceed as normal.
+                registerValidSW(swUrl, config)
+            }
+        })
+        .catch(() => {
+            console.log(
+                'No internet connection found. App is running in offline mode.'
+            )
+        })
+}
+
+export function unregister() {
+    if ('serviceWorker' in navigator) {
+        navigator.serviceWorker.ready
+            .then(registration => {
+                registration.unregister()
+            })
+            .catch(error => {
+                console.error(error.message)
```

Comment:
`navigator.serviceWorker.ready` [never rejects](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerContainer/ready), so any error that's caught would have come from `registration.unregister()` the first time -- I don't think it would be helpful to try unregistering again in that case?